System/Posix/Files/Common.hsc

```diff
@@ -250,6 +251,10 @@ specialDeviceID  :: FileStatus -> DeviceID
 -- | Size of the file in bytes. If this file is a symbolic link the size is
 -- the length of the pathname it contains.
 fileSize         :: FileStatus -> FileOffset
+-- | Number of block blocks allocated for this file, in units of 512-byte.
+fileBlocks       :: FileStatus -> Maybe CBlkCnt
```

Comment:
The types `CBlkCnt` and `CBlkSize` below don't appear to be defined with `GHC <= 8.0` breaking CI.

Reply:
Indeed these were introduced *conditionally* in [GHC 8.2](https://gitlab.haskell.org/ghc/ghc/-/commit/ffc2327070dbb664bdb407a804121eacb2a7c734) and this PR can probably use the corresponding macros to detect their presence.  The new feature can then only be offered if the `stat` structure has the requisite fields (already done) AND the types are available.

Reply:
Ah, right. Not sure how we missed that before. I'll add some more GLASGOW_HASKELL version checking then.

Reply:
This CPP guard is not sufficient. GHC 8.2 [conditionally](https://gitlab.haskell.org/ghc/ghc/-/commit/ffc2327070dbb664bdb407a804121eacb2a7c734#a619a8fa2d426588e918a7b2676a76fac02c51d7_74_74) defines these types.  So you need to:
```
#include "HsBaseConfig.h"
```
and use the same feature macros.

Reply:
Which means that the types might in fact be unsupported on some platforms even with GHC >= 8.2.  However, when that's the case we could define them to be new types for `Void` in that case, and then the functions can return `Maybe X -- X ~ Void`, which can only be instantiated as `Nothing`.

Reply:
I tried that, but that doesn't work because `HsBaseConfig.h` defines a bunch of conflicting defines like PACKAGE_NAME etc. which we also have in `HsUnix.h`.

I'm pretty sure that if the `st_` fields exist the corresponding types must also be defined in the system headers so there shouldn't be any need to check for the types and struct fields separately.

Reply:
> then the functions can return Maybe X -- X ~ Void

I thought about that too, but then the API is conditional depending on platform factors again :)

Reply:
> I'm pretty sure that if the st_ fields exist the corresponding types must also be defined

Looks like these didn't always exist, at least according to https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_stat.h.html

> ### Issue 5
> [...]
> The type of st_blksize is changed from long to blksize_t; the type of st_blocks is changed from long to blkcnt_t.